federatedscope/core/auxiliaries/data_builder.py

```diff
@@ -118,6 +118,238 @@ def _generate_data(client_num=5,
+            test_size = len(dataset) - train_size - val_size
+            lengths = [train_size, val_size, test_size]
+            dataset_train, dataset_val, dataset_test = torch.utils.data.dataset.random_split(
+                dataset, lengths)
+
+        data_dict = {
+            'train': dataset_train,
+            'val': dataset_val,
+            'test': dataset_test
+        }
+
+        return data_dict
+
+    def load_torchtext_data(name, splits=None, config=None):
+        from torch.nn.utils.rnn import pad_sequence
+        from torchtext.data import get_tokenizer
+        from federatedscope.nlp.dataset.utils import label_to_index
+
+        dataset_func = getattr(import_module('torchtext.datasets'), name)
+        if config.data.args:
+            raw_args = config.data.args[0]
+        else:
+            raw_args = {}
+        assert 'max_len' in raw_args, "Miss key 'max_len' in `config.data.args`."
+        filtered_args = filter_dict(dataset_func.__init__, raw_args)
+        dataset = dataset_func(root=config.data.root, **filtered_args)
+
+        # torchtext.transforms requires >= 0.12.0 and torch = 1.11.0,
+        # so we do not use `get_transform` in torchtext.
+        tokenizer = get_tokenizer("basic_english")
+        if len(config.data.transform) == 0:
+            raise ValueError(
+                "`transform` must be one pretrained Word Embeddings from \
+                ['GloVe', 'FastText', 'CharNGram']")
+        if len(config.data.transform) == 1:
+            config.data.transform.append({})
+        vocab = getattr(import_module('torchtext.vocab'),
+                        config.data.transform[0])(dim=config.model.in_channels,
+                                                  **config.data.transform[1])
+        data_list = []
+        for data_iter in dataset:
+            data, targets = [], []
+            if config.model.task == 'seq2seq':
+                for item in data_iter:
+                    data.append(
+                        vocab.get_vecs_by_tokens(tokenizer(item[1]),
+                                                 lower_case_backup=True))
+                    targets.append(
+                        vocab.get_vecs_by_tokens(tokenizer(item[0]),
+                                                 lower_case_backup=True))
+                targets = pad_sequence(targets).transpose(
+                    0, 1)[:, :raw_args['max_len'], :]
+            else:
+                for item in data_iter:
+                    data.append(
+                        vocab.get_vecs_by_tokens(tokenizer(item[1]),
+                                                 lower_case_backup=True))
+                    targets.append(item[0])
+                targets = label_to_index(targets)
+            data = pad_sequence(data).transpose(0,
+                                                1)[:, :raw_args['max_len'], :]
+            data_list.append([(x, y) for x, y in zip(data, targets)])
```

Comment:
For the [IterableDataset](https://pytorch.org/docs/stable/data.html#torch.utils.data.IterableDataset), we may need a more general and principled load function. I will add and test more speech data based on `IterableDataset` and we can improve this in future discussions.

Reply:
Thanks, I will add a `TODO` tag here.